compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -3207,7 +3207,21 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
         case Object_isInstanceOf =>
           genIsAsInstanceOf(obj, targs, cast = false)
         case Object_asInstanceOf =>
-          genIsAsInstanceOf(obj, targs, cast = true)
+          val targetTpe = targs.head.tpe
+          obj match {
+            /* This is an optimization for `linkTimeIf(cond)(thenp)(elsep).asInstanceOf[T]`.
+             * If both `thenp` and `elsep` are subtypes of `T`, the `asInstanceOf`
+             * is redundant and can be removed.
+             */
+            case Apply(fun, List(cond, thenp, elsep))
+                if fun.symbol == jsDefinitions.LinkingInfo_linkTimeIf &&
+                thenp.tpe <:< targetTpe && elsep.tpe <:< targetTpe =>
+              val genObj = genExpr(obj).asInstanceOf[js.LinkTimeIf]
```

Comment:
The original change already looks good to me, but maybe we can match against `genExpr(obj)` and abort if it's not typed `js.LinkTimeIf`? That would display informative error when the compiler generates an unexpected IR by accident.